pyat/at/lattice/transformation.py

```diff
@@ -0,0 +1,280 @@
+    
+    Rotation matrix operator (R1, R2).
+    Can take into account the effect of a longitudinal displacement.
+    ld: Longitudinal displacement [m]
+    r3d: 3D rotation matrix
+    Corresponds to Eq. (9)
+    """
+    return numpy.array([
+        [
+            r3d[1, 1] / r3d[2, 2],
+            ld * r3d[1, 1] / r3d[2, 2] ** 2,
+            -r3d[0, 1] / r3d[2, 2],
+            -ld * r3d[0, 1] / r3d[2, 2] ** 2,
+            0,
+            0,
+        ],
+        [
+            0,
+            r3d[0, 0],
+            0,
+            r3d[1, 0],
+            r3d[2, 0],
+            0,
+        ],
+        [
+            -r3d[1, 0] / r3d[2, 2],
+            -ld * r3d[1, 0] / r3d[2, 2] ** 2,
+            r3d[0, 0] / r3d[2, 2],
+            ld * r3d[0, 0] / r3d[2, 2] ** 2,
+            0,
+            0,
+        ],
+        [
+            0,
+            r3d[0, 1],
+            0,
+            r3d[1, 1],
+            r3d[2, 1],
+            0,
+        ],
+        [
+            0,
+            0,
+            0,
+            0,
+            1,
+            0,
+        ],
+        [
+            -r3d[0, 2] / r3d[2, 2],
+            -ld * r3d[0, 2] / r3d[2, 2] ** 2,
+            -r3d[1, 2] / r3d[2, 2],
+            -ld * r3d[1, 2] / r3d[2, 2] ** 2,
+            0,
+            1,
+        ],
+    ])
+    
+def transform_elem(elem: Element, midpoint: str = "center",
+                   dx: float = 0.0, dy: float = 0.0,                       
+                   tilt: float = 0.0, pitch: float = 0.0, yaw: float = 0.0,
+                   relative: bool = False) -> None:
+    r"""Set the tilt, pitch and yaw angle of an :py:class:`.Element`.
+    The tilt is a rotation around the *s*-axis, the pitch is a
+    rotation around the *x*-axis and the yaw is a rotation around
+    the *y*-axis.
+
+    A positive angle represents a clockwise rotation when
+    looking in the direction of the rotation axis.
+
+    The transformations are not all commmutative. The rotations are applied in 
+    the order *Z* -> *Y* -> *X* (tilt -> yaw -> pitch). The element is rotated 
+    around its mid-point. The mid-point can either be the element entrance 
+    (entry face of the downstream drift element) or center (axis joining the 
+    entry and exit points of the element).
+
+    If *relative* is :py:obj:`True`, the previous angles are rebuilt from the 
+    *r3d* matrix and incremented by the input arguments.
+    *relative* only allows to add the previous angles, not the transverse 
+    shifts.
+    The shift is always conserved regardless of the value of *relative*.
+    
+    pyAT describes the ultra-relativistic beam dynamics in 6D phase space 
+    coordinates, which differ from 3D spatial angles in an expansion with 
+    respect to the energy to first order by a factor (1 + $\\delta$) , where 
+    $\\delta$ is the relative energy offset. In general this introduces a small 
+    angle error, but could create an undesired effect for large energy offsets.
```

Comment:
It introduces a (hopefully small)  **spurious dispersion**: it's an angle proportional to the off-momentum. I would rephrase it that way…